File: app/src/main/java/com/android/example/popularmovies/ui/detail/fragments/reviews/ReviewsViewModel.java
package com.android.example.popularmovies.ui.detail.fragments.reviews;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.android.example.popularmovies.AppExecutors;
import com.android.example.popularmovies.utils.json.MovieReview;
import com.android.example.popularmovies.utils.network.LoadingStatus;
import com.android.example.popularmovies.utils.network.NetworkUtils;

import java.util.List;

public class ReviewsViewModel extends ViewModel {

    private MutableLiveData<LoadingStatus> status;
    private MutableLiveData<List<MovieReview>> reviews;

    LiveData<LoadingStatus> getStatus(){
        return status;
    }

    @SuppressWarnings("SameParameterValue")
    void setStatus(LoadingStatus newStatus) {
        status.postValue(newStatus);
    }

    public LiveData<List<MovieReview>> getReviews(){
        return reviews;
    }

    public ReviewsViewModel() {
        super();
        status = new MutableLiveData<>();
        reviews = new MutableLiveData<>();
        status.postValue(LoadingStatus.INIT);
    }

    void fetchReviews(@NonNull String apiKey, int movieId){
        status.postValue(LoadingStatus.LOADING);
        AppExecutors.getInstance().networkIO().execute(() -> {
            List<MovieReview> fetchedReviews = NetworkUtils.getMovieReviews(movieId, apiKey);
            if (fetchedReviews != null){
                reviews.postValue(fetchedReviews);
                status.postValue(LoadingStatus.DONE);
            } else {
                status.postValue(LoadingStatus.ERROR);
            }
        });
    }
}

File: app/src/main/java/com/android/example/popularmovies/ui/detail/fragments/reviews/ReviewsViewModel.kt
package com.android.example.popularmovies.ui.detail.fragments.reviews

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.android.example.popularmovies.utils.json.MovieReview
import com.android.example.popularmovies.utils.network.LoadingStatus
import com.android.example.popularmovies.utils.network.MovieDbAPI
import kotlinx.coroutines.launch

class ReviewsViewModel : ViewModel() {
    private val _status: MutableLiveData<LoadingStatus> = MutableLiveData()
    val status: LiveData<LoadingStatus> get() = _status
    private val _reviews: MutableLiveData<List<MovieReview>> = MutableLiveData()
    val reviews: LiveData<List<MovieReview>> get() = _reviews

    fun setStatus(newStatus: LoadingStatus) {
        _status.postValue(newStatus)
    }

    fun fetchReviews(apiKey: String, movieId: Int) {
        _status.postValue(LoadingStatus.LOADING)
        viewModelScope.launch {
            try {
                _reviews.postValue(MovieDbAPI.apiService.getMovieReviews(movieId, apiKey))
                _status.postValue(LoadingStatus.DONE)
            } catch (e: Exception) {
                _status.postValue(LoadingStatus.ERROR)
            }
        }
    }

    init {
        _status.postValue(LoadingStatus.INIT)
    }
}